### PR TITLE
[wasm] Mark System.IO.FileSystem.Watcher APIs as unsupported on Browser

### PR DESCRIPTION
**Pull request by mdh1418**
Contributes to #41087 

```
M:System.IO.ErrorEventArgs.#ctor(System.Exception),System.IO,ErrorEventArgs,.ctor(Exception),0
M:System.IO.ErrorEventArgs.GetException,System.IO,ErrorEventArgs,GetException(),0
"M:System.IO.FileSystemEventArgs.#ctor(System.IO.WatcherChangeTypes,System.String,System.String)",System.IO,FileSystemEventArgs,".ctor(WatcherChangeTypes, String, String)",0
M:System.IO.FileSystemEventArgs.get_ChangeType,System.IO,FileSystemEventArgs,get_ChangeType(),0
M:System.IO.FileSystemEventArgs.get_Name,System.IO,FileSystemEventArgs,get_Name(),0
M:System.IO.FileSystemEventArgs.get_FullPath,System.IO,FileSystemEventArgs,get_FullPath(),0
M:System.IO.FileSystemWatcher.get_NotifyFilter,System.IO,FileSystemWatcher,get_NotifyFilter(),0
M:System.IO.FileSystemWatcher.set_NotifyFilter(System.IO.NotifyFilters),System.IO,FileSystemWatcher,set_NotifyFilter(NotifyFilters),0
M:System.IO.FileSystemWatcher.get_Path,System.IO,FileSystemWatcher,get_Path(),0
M:System.IO.FileSystemWatcher.get_SynchronizingObject,System.IO,FileSystemWatcher,get_SynchronizingObject(),0
M:System.IO.FileSystemWatcher.BeginInit,System.IO,FileSystemWatcher,BeginInit(),0
M:System.IO.FileSystemWatcher.get_Site,System.IO,FileSystemWatcher,get_Site(),0
M:System.IO.FileSystemWatcher.WaitForChanged(System.IO.WatcherChangeTypes),System.IO,FileSystemWatcher,WaitForChanged(WatcherChangeTypes),0
M:System.IO.FileSystemWatcher.OnRenamed(System.IO.RenamedEventArgs),System.IO,FileSystemWatcher,OnRenamed(RenamedEventArgs),0
M:System.IO.FileSystemWatcher.OnError(System.IO.ErrorEventArgs),System.IO,FileSystemWatcher,OnError(ErrorEventArgs),0
M:System.IO.FileSystemWatcher.get_EnableRaisingEvents,System.IO,FileSystemWatcher,get_EnableRaisingEvents(),0
M:System.IO.FileSystemWatcher.set_Path(System.String),System.IO,FileSystemWatcher,set_Path(String),0
M:System.IO.FileSystemWatcher.set_InternalBufferSize(System.Int32),System.IO,FileSystemWatcher,set_InternalBufferSize(Int32),0
M:System.IO.FileSystemWatcher.get_IncludeSubdirectories,System.IO,FileSystemWatcher,get_IncludeSubdirectories(),0
M:System.IO.FileSystemWatcher.OnCreated(System.IO.FileSystemEventArgs),System.IO,FileSystemWatcher,OnCreated(FileSystemEventArgs),0
M:System.IO.FileSystemWatcher.OnDeleted(System.IO.FileSystemEventArgs),System.IO,FileSystemWatcher,OnDeleted(FileSystemEventArgs),0
"M:System.IO.FileSystemWatcher.WaitForChanged(System.IO.WatcherChangeTypes,System.Int32)",System.IO,FileSystemWatcher,"WaitForChanged(WatcherChangeTypes, Int32)",0
M:System.IO.FileSystemWatcher.set_SynchronizingObject(System.ComponentModel.ISynchronizeInvoke),System.IO,FileSystemWatcher,set_SynchronizingObject(ISynchronizeInvoke),0
M:System.IO.FileSystemWatcher.#ctor,System.IO,FileSystemWatcher,.ctor(),0
M:System.IO.FileSystemWatcher.set_IncludeSubdirectories(System.Boolean),System.IO,FileSystemWatcher,set_IncludeSubdirectories(Boolean),0
M:System.IO.FileSystemWatcher.OnChanged(System.IO.FileSystemEventArgs),System.IO,FileSystemWatcher,OnChanged(FileSystemEventArgs),0
M:System.IO.FileSystemWatcher.get_InternalBufferSize,System.IO,FileSystemWatcher,get_InternalBufferSize(),0
M:System.IO.FileSystemWatcher.set_EnableRaisingEvents(System.Boolean),System.IO,FileSystemWatcher,set_EnableRaisingEvents(Boolean),0
M:System.IO.FileSystemWatcher.set_Filter(System.String),System.IO,FileSystemWatcher,set_Filter(String),0
"M:System.IO.FileSystemWatcher.#ctor(System.String,System.String)",System.IO,FileSystemWatcher,".ctor(String, String)",0
M:System.IO.FileSystemWatcher.set_Site(System.ComponentModel.ISite),System.IO,FileSystemWatcher,set_Site(ISite),0
M:System.IO.FileSystemWatcher.#ctor(System.String),System.IO,FileSystemWatcher,.ctor(String),0
M:System.IO.FileSystemWatcher.get_Filter,System.IO,FileSystemWatcher,get_Filter(),0
M:System.IO.FileSystemWatcher.EndInit,System.IO,FileSystemWatcher,EndInit(),0
M:System.IO.FileSystemWatcher.get_Filters,System.IO,FileSystemWatcher,get_Filters(),0
"M:System.IO.InternalBufferOverflowException.#ctor(System.String,System.Exception)",System.IO,InternalBufferOverflowException,".ctor(String, Exception)",0
M:System.IO.InternalBufferOverflowException.#ctor(System.String),System.IO,InternalBufferOverflowException,.ctor(String),0
"M:System.IO.InternalBufferOverflowException.#ctor(System.Runtime.Serialization.SerializationInfo,System.Runtime.Serialization.StreamingContext)",System.IO,InternalBufferOverflowException,".ctor(SerializationInfo, StreamingContext)",0
M:System.IO.InternalBufferOverflowException.#ctor,System.IO,InternalBufferOverflowException,.ctor(),0
"M:System.IO.RenamedEventArgs.#ctor(System.IO.WatcherChangeTypes,System.String,System.String,System.String)",System.IO,RenamedEventArgs,".ctor(WatcherChangeTypes, String, String, String)",0
M:System.IO.RenamedEventArgs.get_OldName,System.IO,RenamedEventArgs,get_OldName(),0
M:System.IO.RenamedEventArgs.get_OldFullPath,System.IO,RenamedEventArgs,get_OldFullPath(),0
M:System.IO.WaitForChangedResult.set_TimedOut(System.Boolean),System.IO,WaitForChangedResult,set_TimedOut(Boolean),0
M:System.IO.WaitForChangedResult.get_ChangeType,System.IO,WaitForChangedResult,get_ChangeType(),0
M:System.IO.WaitForChangedResult.set_OldName(System.String),System.IO,WaitForChangedResult,set_OldName(String),0
M:System.IO.WaitForChangedResult.set_ChangeType(System.IO.WatcherChangeTypes),System.IO,WaitForChangedResult,set_ChangeType(WatcherChangeTypes),0
M:System.IO.WaitForChangedResult.get_TimedOut,System.IO,WaitForChangedResult,get_TimedOut(),0
M:System.IO.WaitForChangedResult.get_OldName,System.IO,WaitForChangedResult,get_OldName(),0
M:System.IO.WaitForChangedResult.set_Name(System.String),System.IO,WaitForChangedResult,set_Name(String),0
M:System.IO.WaitForChangedResult.get_Name,System.IO,WaitForChangedResult,get_Name(),0
```

### CODE REVIEW

File: src/libraries/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs [hunk 1]
```diff
@@ -72,18 +103,25 @@ public enum NotifyFilters
...
         public string OldFullPath { get { throw null; } }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         public string? OldName { get { throw null; } }
     }
     public delegate void RenamedEventHandler(object sender, System.IO.RenamedEventArgs e);
     public partial struct WaitForChangedResult
     {
         private object _dummy;
         private int _dummyPrimitive;
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
```
Comment (marek-safar) [inline src/libraries/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs:118]: Why is it not on class?

File: src/libraries/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs [hunk 2]
```diff
@@ -72,18 +103,25 @@ public enum NotifyFilters
     }
     public partial class RenamedEventArgs : System.IO.FileSystemEventArgs
     {
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
```
Comment (marek-safar) [inline src/libraries/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs:106]: Why is it not on class?

File: src/libraries/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs [hunk 3]
```diff
@@ -8,54 +8,85 @@ namespace System.IO
...
         public ErrorEventArgs(System.Exception exception) { }
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
         public virtual System.Exception GetException() { throw null; }
     }
     public delegate void ErrorEventHandler(object sender, System.IO.ErrorEventArgs e);
     public partial class FileSystemEventArgs : System.EventArgs
     {
+        [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
```
Comment (marek-safar) [inline src/libraries/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs:19]: Why is it not on class?

**Comment by mdh1418** [comment]
Changed to mark UnsupportedOSPlatforms at assembly level since it is skipped for tests at the assembly level. https://github.com/dotnet/runtime/blob/740d59ab92806533a5116cdec2a00f634f568264/src/libraries/System.IO.FileSystem.Watcher/tests/AssemblyInfo.cs#L7

File: src/libraries/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs (hunk 1 above)
Comment (mdh1418) [reply src/libraries/System.IO.FileSystem.Watcher/ref/System.IO.FileSystem.Watcher.cs]: Right, it should be. I now realize the whole assembly can be skipped https://github.com/dotnet/runtime/pull/41682#issuecomment-687255619

**Review by marek-safar — APPROVED** [review]

**Review by akoeplinger — APPROVED** [review]

**Pull request merged**